### PR TITLE
Lock down write access to company's org members and provide anonymous public access

### PR DESCRIPTION
**Pull request by belucid**
https://trello.com/c/nOF9AoKg

This PR BOTH opens up access to anonymous API users w/o a JWToken AND locks down access to only a company's organization. Before we required a JWToken for everything, but any valid JWToken would do for any request. Now it's more complicated, requests are either:

OK for anonymous:
- invalid JWToken **not** allowed
- **no** JWToken allowed
- any valid JWToken allowed

OK only for company org. members:
- invalid JWToken **not** allowed
- **no** JWToken **not** allowed
- valid JWToken for another org **not** allowed
- valid JWToken for company's org allowed

This spreadsheet details the security rules and expected response for every API call in use at this point:
https://docs.google.com/spreadsheets/d/1AvDqPHzDTWUxUdpzw9Hlff4IWbxc2QwHvcwGbVrYoSg/edit#gid=0

This PR implements the rules and responses of that spreadsheet, and provides tests for the same.

To test:
0. Review the code changes
1. Review the tests, compare them to the [spreadsheet](https://docs.google.com/spreadsheets/d/1AvDqPHzDTWUxUdpzw9Hlff4IWbxc2QwHvcwGbVrYoSg/edit#gid=0). Do they adequately test everything there?
2. Run the tests with `lein test!`
3. Reimport all the sample company data (it now stores the organization)
4. Double check any unit tests you want to using the REPL
5. Double check any integration tests you want to using cURL and the client UI
6. For the client UI, **remove your JWToken cookies first**, and it'll be handy to test with the company `read-only`, take a look at the company GET request, it has no HATEOAS links except for 'self' on company and sections? Take a look at the OPTIONS request for `sections/new`, it 401's? Groovy!
7. Merge
8. Deploy to staging/prod
9. Reimport all sample companies staging/prod
10. Sing the Star Spangled Banner a cappella or with ukulele accompaniment (no guitar!)

### CODE REVIEW

**Comment by belucid** [comment]
@bago2k4 this is just a placeholder for me right now. Still working on a few final tests and tweaks. Don't review yet.

**Comment by belucid** [comment]
@bago2k4 This is ready now.

File: test/open_company/integration/company/company_list.clj [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  (facts "about listing companies"
+
+    (fact "all existing companies are listed anonymously with NO JWToken"
+      (let [response (mock/api-request :get "/companies" {:skip-auth true})
+            body (mock/body-from-response response)
+            companies (get-in body [:collection :companies])]
+        (:status response) => 200
+        (map :name companies) => (just (set (map :name [r/open r/uni r/buffer]))) ; verify names
+        (map :name companies) => (just (set (map :name [r/open r/uni r/buffer]))) ; verify names
```
Comment (bago2k4) [inline test/open_company/integration/company/company_list.clj]: this is a duplicated line

File: test/open_company/integration/company/company_retrieval.clj [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    (fact "that doesn't match the retrieving user's organization"
+      ;; retrieve with Sartre (different org)
+      (let [response (mock/api-request :get (company-rep/url r/open) {:auth mock/jwtoken-sartre})
+            body (mock/body-from-response response)]
+        (:status response) => 200
+        (:sections body) =>
+          {:company ["diversity" "values"], :financial ["finances"], :progress ["update" "team" "help"]}
+        ;; verify each section has only a self HATEOAS link
+        (doseq [section-key (map keyword (flatten (vals (:sections body))))]
+          (count (:links (body section-key))) => 1)
```
Comment (bago2k4) [inline test/open_company/integration/company/company_retrieval.clj]: @belucid probably better to use `verify-link` here too

File: src/open_company/resources/common.clj [hunk 3]
```diff
@@ -73,6 +74,21 @@
...
+  (if (s/blank? (:real-name user))
+    user
+    (assoc user :name (:real-name user))))
+
+(defn author-for-user
+  "Extract the :avatar/:image, :user-id and :name (the author fields) from the JWToken claims."
+  [user]
+  (-> user
+    (name-for)
+    (select-keys [:avatar :user-id :name])
```
Comment (bago2k4) [inline src/open_company/resources/common.clj]: @belucid i think the problem with the author images is here, changing this line with this
```
(select-keys [:avatar :user-id :name :image])
```
and reimporting all the companies worked for me. I leave the change to you to be sure it's the right fix.

**Comment by bago2k4** [comment]
@belucid above comments apart this looks good. you can merge and deploy if you want.
Problem is: if you merge and deploy this who's to sing and play Star Spangled Banner? i don't have an ukulele..
Marry Christmas!

File: test/open_company/integration/company/company_list.clj (hunk 1 above)
Comment (belucid) [inline test/open_company/integration/company/company_list.clj]: Good catch @bago2k4

File: test/open_company/integration/company/company_retrieval.clj (hunk 2 above)
Comment (belucid) [inline test/open_company/integration/company/company_retrieval.clj]: agreed @bago2k4

**Pull request merged**